### PR TITLE
added translational invariant kernel class, refactored distance kernel

### PR DESCRIPTION
**Pull request by lambday**
@karlnapf hey! Please have a look. Will send the gaussian kernel refactoring in another patch.

### CODE REVIEW

File: src/shogun/kernel/BesselKernel.cpp [hunk 1]
```diff
@@ -23,8 +23,6 @@ CBesselKernel::CBesselKernel(int32_t size, float64_t v, float64_t w,
 		int32_t n, CDistance* dist) :
 	CDistanceKernel(size,w,dist), order(v), degree(n)
 {
-	ASSERT(distance)
```
Comment (karlnapf) [inline src/shogun/kernel/BesselKernel.cpp]: ?

**Comment by karlnapf** [comment]
Can I suggest we split this PR?
Have the base class first (addition only, unit tests), then refactor other kernels one by one
I will ignore all other files than translation invariant for now if that is ok. If not, let me know

File: src/shogun/kernel/TranslationalInvariantKernel.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+#ifndef TRANSLATIONAL_INVARIANT_KERNEL_H_
+#define TRANSLATIONAL_INVARIANT_KERNEL_H_
+
+#include <shogun/kernel/Kernel.h>
+#include <shogun/distance/Distance.h>
+
+namespace shogun
+{
+
+/** @brief The translational invariant kernel */
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.h]: This should have some math and the statement that this is an abstract base class

File: src/shogun/kernel/TranslationalInvariantKernel.h [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+public:
+	/** Default constructor.  */
+	CTranslationalInvariantKernel();
+
+	/**
+	 * Constructor with a specified cache size.
+	 *
+	 * @param cache cache size
+	 */
+	CTranslationalInvariantKernel(int32_t cache);
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.h]: I guess we dont need this....

File: src/shogun/kernel/TranslationalInvariantKernel.h [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+	/**
+	 * Constructor that initializes the kernel with two feature instances.
+	 *
+	 * @param l features of left-hand side
+	 * @param r features of right-hand side
+	 * @param cache cache size
+	 */
+	CTranslationalInvariantKernel(CFeatures *l, CFeatures *r, int32_t cache);
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.h]: I would do the cache solely via setter

File: src/shogun/kernel/TranslationalInvariantKernel.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+CTranslationalInvariantKernel::~CTranslationalInvariantKernel()
+{
+	cleanup();
+	SG_UNREF(m_distance);
+}
+
+bool CTranslationalInvariantKernel::init(CFeatures* l, CFeatures* r)
+{
+	CKernel::init(l,r);
+	if (m_distance!=NULL)
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: would this call even be allowed without a distance set? Does it have a meaning?

File: src/shogun/kernel/TranslationalInvariantKernel.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		m_distance->init(l, r);
+	return init_normalizer();
+}
+
+void CTranslationalInvariantKernel::set_distance(CDistance* distance)
+{
+	ASSERT(distance);
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: REQUIRE

File: src/shogun/kernel/TranslationalInvariantKernel.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+void CTranslationalInvariantKernel::register_params()
+{
+	SG_ADD((CSGObject**) &m_distance, "m_distance", "Distance to be used.", MS_AVAILABLE);
+}
+
+void CTranslationalInvariantKernel::precompute_distance()
+{
+	REQUIRE(m_distance, "The distance instance is not initialized! Please provide the distance with set_distance method.\n");
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: Maybe not mention the method here (but just THAT distance has to be providced) as these things tend to change.

File: src/shogun/kernel/TranslationalInvariantKernel.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	REQUIRE(m_distance->init(lhs, rhs), "Could not initialize the distance instance! Please contact Shogun developers.\n");
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: Remove "Please contact"

File: src/shogun/kernel/TranslationalInvariantKernel.cpp [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	try
+	{
+		SGMatrix<float32_t> dist_mat=m_distance->get_distance_matrix<float32_t>();
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: why compute the full matrix if only triangular is used?

File: src/shogun/kernel/TranslationalInvariantKernel.cpp [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+		CCustomDistance* distance=new CCustomDistance();
+		SG_REF(distance);
+		distance->set_triangle_distance_matrix_from_full(dist_mat.data(), dist_mat.num_rows, dist_mat.num_cols);
+		SG_UNREF(m_distance);
+		m_distance=distance;
+	}
+	catch (ShogunException e)
+	{
+		SG_ERROR("Pre-computing distance was not possible! Please contact Shogun developers.\n");
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: not sure about this....any error will stop execution anyways

File: src/shogun/kernel/TranslationalInvariantKernel.cpp (hunk 5 above)
Comment (lambday) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: If a subclass (say, GaussianKernel) does not initialize a distance instance, then also it should work, no? For CDistanceKernel, this instance is mandatory.

File: src/shogun/kernel/TranslationalInvariantKernel.cpp (hunk 9 above)
Comment (lambday) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: Yeah triangular should be better. But the distance API currently doesn't have that.

Also, this is a bug. I should only use triangular matrix when lhs and rhs are equal. Fixing it.

File: src/shogun/kernel/TranslationalInvariantKernel.h (hunk 4 above)
Comment (lambday) [inline src/shogun/kernel/TranslationalInvariantKernel.h]: yeah that's better. This is just to imitate the API we have for CKernel. I'll change it.

**Comment by lambday** [comment]
@karlnapf updated the PR. Just sent the new class with its unit-tests for now. Please have a look!

File: src/shogun/kernel/TranslationalInvariantKernel.cpp (hunk 5 above)
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: you are right +1

File: src/shogun/kernel/TranslationalInvariantKernel.cpp [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+CDistance* CTranslationalInvariantKernel::get_distance() const
+{
+	SG_REF(m_distance);
+	return m_distance;
+}
+
+void CTranslationalInvariantKernel::precompute_distance()
+{
+	REQUIRE(m_distance, "The distance instance is not initialized!"
+			"Please provide the distance with set_distance method.\n");
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: can we remove the setter method name in the message here?

Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: "The distance instance is not initialized, please provide one."

File: src/shogun/kernel/TranslationalInvariantKernel.cpp [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	REQUIRE(m_distance->init(lhs, rhs), "Could not initialize the distance instance!\n");
+
+	SGMatrix<float32_t> dist_mat=m_distance->get_distance_matrix<float32_t>();
+	CCustomDistance* distance=new CCustomDistance();
+
+	if (lhs==rhs)
+		distance->set_triangle_distance_matrix_from_full(dist_mat.data(), dist_mat.num_rows, dist_mat.num_cols);
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: yep, phew :)
Is this unit tested?

File: src/shogun/kernel/TranslationalInvariantKernel.cpp [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+	else
+		distance->set_full_distance_matrix_from_full(dist_mat.data(), dist_mat.num_rows, dist_mat.num_cols);
+
+	set_distance(distance);
+}
+
+void CTranslationalInvariantKernel::register_params()
+{
+	SG_ADD((CSGObject**) &m_distance, "m_distance", "Distance to be used.", MS_AVAILABLE);
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.cpp]: No model selection here pls

File: src/shogun/kernel/TranslationalInvariantKernel.h [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+#ifndef TRANSLATIONAL_INVARIANT_KERNEL_H_
+#define TRANSLATIONAL_INVARIANT_KERNEL_H_
+
+#include <shogun/kernel/Kernel.h>
+#include <shogun/distance/Distance.h>
+
+namespace shogun
+{
+
+/** @brief The translational invariant kernel class that represents the base
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.h]: translation invariant

Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.h]: Start like this:
"Base class that represents kernel functions that only depend on the difference of its inputs, i.e."

File: src/shogun/kernel/TranslationalInvariantKernel.h [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+ * class of the kernel functions whose values don't change if the feature
+ * vectors are shifted by an equal amount. Mathematically,
+ * \f[
+ * 	k(\mathbf{x}, \mathbf{x'}) = k(\mathbf{x+u}, \mathbf{x'+u})
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.h]: would rather say. The kernel can be written as
k(x,y) = k(x-y)

File: src/shogun/kernel/TranslationalInvariantKernel.h [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+	/** @return feature class of distance used */
+	virtual EFeatureClass get_feature_class() = 0;
+
+	/** @return the distance type */
+	virtual EDistanceType get_distance_type() = 0;
+
+	/** @return name Distance */
+	virtual const char* get_name() const
+	{
+		return "TranslationalInvariantKernel";
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationalInvariantKernel.h]: TranslationInvariantKernel is a better name

File: tests/unit/kernel/TranslationalInvariantKernel_unittest.cc [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+#include <algorithm>
+
+using namespace shogun;
+using std::iota;
+using std::for_each;
+
+class CTranslationalInvariantKernelMock : public CTranslationalInvariantKernel
+{
+public:
+	MOCK_METHOD2(compute, float64_t(int32_t, int32_t));
```
Comment (karlnapf) [inline tests/unit/kernel/TranslationalInvariantKernel_unittest.cc]: googlemock for the rescue! :)

File: tests/unit/kernel/TranslationalInvariantKernel_unittest.cc [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+	for_each(data_1.data(), data_1.data()+data_1.size(), [&data_1](float64_t& val) { val=val/data_1.size(); });
+	for_each(data_2.data(), data_2.data()+data_2.size(), [&data_2](float64_t& val) { val=val/data_2.size(); });
+
+	auto feats_1=new CDenseFeatures<float64_t>(data_1);
+	auto feats_2=new CDenseFeatures<float64_t>(data_2);
+
+	auto euclidean_distance=new CEuclideanDistance();
+	SG_REF(euclidean_distance);
+
+	auto kernel=some<CTranslationalInvariantKernelMock>();
```
Comment (karlnapf) [inline tests/unit/kernel/TranslationalInvariantKernel_unittest.cc]: any reason to mix some with new?

File: tests/unit/kernel/TranslationalInvariantKernel_unittest.cc [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+
+	euclidean_distance->init(feats_1, feats_2);
+
+	for (auto i=0; i<N; ++i)
+	{
+		for (auto j=0; j<M; ++j)
+			EXPECT_NEAR(euclidean_distance->distance(i, j), precomputed_distance->distance(i, j), 1E-6);
+	}
+
+	SG_UNREF(euclidean_distance);
```
Comment (karlnapf) [inline tests/unit/kernel/TranslationalInvariantKernel_unittest.cc]: some can avoid these unrefs

**Comment by karlnapf** [comment]
Fine to merge from my side apart from some minor changes requires, the most important one being the class name change...

**Comment by lambday** [comment]
@karlnapf updated. could you please have a quick look and merge if things look okay?

**Comment by lisitsyn** [comment]
Would shift invariant be a shorter name for the same thing?

**Comment by karlnapf** [comment]
Fine to merge otherwise

File: src/shogun/kernel/TranslationInvariantKernel.h [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+
+namespace shogun
+{
+
+/** @brief Base class for the family of kernel functions that only depend on
+ * the difference of the inputs. More precisely,
+ * \f[
+ * 	k(\mathbf{x}, \mathbf{x'}) = k(\mathbf{x-x'})
+ * \f]
+ * For example, Gaussian (RBF) kernel is a translational invariant kernel.
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationInvariantKernel.h]: shift

File: src/shogun/kernel/TranslationInvariantKernel.h [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+#define TRANSLATION_INVARIANT_KERNEL_H_
+
+#include <shogun/kernel/Kernel.h>
+#include <shogun/distance/Distance.h>
+
+namespace shogun
+{
+
+/** @brief Base class for the family of kernel functions that only depend on
+ * the difference of the inputs. More precisely,
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationInvariantKernel.h]: , i.e. whose values does not change if the inputs are shifted by the same amount

File: src/shogun/kernel/TranslationInvariantKernel.h [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+	 * @return if initializing was successful
+	 */
+	virtual bool init(CFeatures* l, CFeatures* r);
+
+	/**
+	 * Method that sets the distance.
+	 *
+	 * @param distance The distance object.
+	 */
+	void set_distance(CDistance* distance);
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationInvariantKernel.h]: remove

File: src/shogun/kernel/TranslationInvariantKernel.h [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+	 * of the feature vectors in the corresponding feature object.
+	 *
+	 * @param idx_a index a
+	 * @param idx_b index b
+	 * @return distance between features a and b
+	 */
+	virtual float64_t distance(int32_t idx_a, int32_t idx_b) = 0;
+
+	/** Distance */
+	CDistance* m_distance;
```
Comment (karlnapf) [inline src/shogun/kernel/TranslationInvariantKernel.h]: CustomDistance

**Pull request merged**